File: go/0028E-实现strStr().go
package main

import "strings"

func getNext(next []int, s string) {
	i, j := 0, -1
	next[0] = j
	for i < len(s) {
		for j !=-1 && s[i] != s[j] { // 前后缀不同，向前回退
			j = next[j]
		}
		i++
		j++
		if i >= len(s) || j >= len(s) {
			return
		}
		if s[i] == s[j] { // 优化去重
			next[i] = next[j]
		} else {
			next[i] = j
		}
	}
}

func strStr(haystack string, needle string) int {
	if len(needle) == 0 {
		return 0
	}
	next := make([]int, len(needle))
	getNext(next, needle)
	i, j := 0, 0
	for i < len(haystack) && j < len(needle) {
		for j !=-1 && haystack[i] != needle[j] {
			j = next[j]
		}
		i++
		j++
	}
	if j == len(needle) {
		return i - j
	} else {
		return -1
	}
}

func strStrOld(haystack string, needle string) int {
	return strings.Index(haystack, needle)
}

File: cpp/0028E-实现strStr().cpp
/*
 * @Author: SukiEva
 * @Date: 2021-12-16 15:22:12
 * @Description: https://github.com/SukiEva
 * わたし、気になります！
 */
#include <bits/stdc++.h>

#include "include/ListNode.h"
#include "include/TreeNode.h"

using namespace std;

class Solution {
   public:
    int strStr(string haystack, string needle) {
        int i = 0, j = 0;
        int nx[needle.size() + 1];
        getNext(needle, nx);
        while (i < haystack.size() && j < needle.size()) {
            while (-1 != j && haystack[i] != needle[j]) j = nx[j];
            i++, j++;
        }
        if (j == needle.size()) return i - j;
        return -1;
    }

    void getNext(const string& s, int* nx) {
        int i = 0;
        int j = nx[0] = -1;
        while (i < s.size()) {
            while (-1 != j && s[i] != s[j]) j = nx[j];
            if (s[++i] == s[++j])
                nx[i] = nx[j];
            else
                nx[i] = j;
        }
    }
};